packages/roosterjs-content-model-plugins/lib/tableEdit/editors/features/TableMover.ts

```diff
@@ -129,3 +157,214 @@ function isTableTopVisible(editor: IEditor, rect: Rect | null, contentDiv?: Node
+            // Move table to new position
+            finalRange.range.collapse(true);
+            let insertionSuccess: boolean = false;
+
+            // Insert table into content model
+            editor.formatContentModel(
+                (model, context) => {
+                    const SPArray = getSelectedSegmentsAndParagraphs(model, false);
+                    if (
+                        SPArray.length == 1 &&
+                        SPArray[0][0].segmentType == 'SelectionMarker' &&
+                        SPArray[0][1] != null
+                    ) {
+                        const paragraph = SPArray[0][1];
+                        paragraph.segments.forEach(segment => {
+                            if (segment == SPArray[0][0] && initValue.cmTable) {
+                                const doc = createContentModelDocument();
+                                doc.blocks.push(initValue.cmTable);
+                                insertionSuccess = !!mergeModel(model, doc, context, {
+                                    mergeFormat: 'none',
+                                });
+                                context.skipUndoSnapshot = true;
+                                // Add selection marker to the first cell of the table
+                                const [finalTable] = getFirstSelectedTable(model);
+
+                                if (finalTable) {
+                                    setSelection(model);
+                                    const FirstCell = finalTable.rows[0].cells[0];
+
+                                    const markerParagraph = FirstCell?.blocks[0];
+                                    if (markerParagraph?.blockType == 'Paragraph') {
+                                        const marker = createSelectionMarker(model.format);
+
+                                        markerParagraph.segments.unshift(marker);
+                                        setParagraphNotImplicit(markerParagraph);
+                                        setSelection(FirstCell, marker);
+                                    }
+                                } else {
+                                    // Restore initial selection
+                                    editor.setDOMSelection(initValue.initialSelection);
+                                }
+                            }
+                        });
+                    }
+                    return true;
+                },
+                {
+                    apiName: 'TableMover',
+                    selectionOverride: {
+                        type: 'range',
+                        range: finalRange.range,
+                        isReverted: false,
+                    },
+                }
+            );
+            // Remove old table only if insertion was successful
+            if (insertionSuccess) {
+                if (!(event.ctrlKey || event.metaKey)) {
+                    table.remove();
```

Comment:
This is risky since the original table element can be overwritten when write back

Reply:
I think we can use `formatInsertPointWithContentModel` to help do the merging. It allows pass in an additional insert point to mark the model together with the existing one.

For example:

```ts
const domTable = ... // The HTML Table that is dragging
const domIP = getNodePositionFromEvent(editor, x, y); // Get the insert point

formatInsertPointWithContentModel(editor, (model,context,ip) => {
  const [table, path] = getFirstSelectedTable(model);
  const index = path[0].blocks.indexOf(table);
  // Add necessary null check here
  path[0].blocks.splice(index, 1); // Table is removed here

 ip.marker.isSelected = true // ip is the insert point generated from the addtional selection passed below, by default it is marked as isSelected =false, we set it to true here to make it a regular selection marker
 
 const newTableModel = ....
 mergeModel(model, newTableModel); // Now this will be merged into the marker we marked above
},
 domIP, // Pass in the target ip here, it will be marked as ip with isSelected= false
 {
   selectionOverride: { // Override the original selection with the dragging table, so it we can use getFirstSelectedTable to get this table in content model
     type: 'table',
     table: domTable,
     ...
   }
});
 
```

Reply:
> This is risky since the original table element can be overwritten when write back

I need more information about this, because I have no encountered any overwriting issues.

Reply:
For example:
```ts
const element = document.querySelector(...) // or any other way that you get a HTML element from editor

editor.formatContentModel( .... );

anyOperationOn(element); // Risky here
```

Because after formatContentModel() is called, it is possible the whole document is regenerated, so it is possible the original element does not exist in editor any more.

Today it works for you possibly because we treat the table that is not in selection as a fake entity, so the TABLE element will be preserved. But this behavior can be changed in the future. We treat table as entity just because of content fidelity issue. Given most of the known content fidelity bug is already fixed, we may remove this hacky behavior. So that any other feature should not take dependency to this behavior.

In general, with content model, we should avoid accessing DOM tree directly as much as possible. We should only access DOM when handling DOM event (in which case we only get DOM info from the event) or for performance reason. Content Model already provides ways to do this without accessing DOM tree after formatContentModel.